### PR TITLE
feat(parsers): use gdamore/tree-sitter-d parser for D

### PR DESCRIPTION
**Pull request by ljmf00**
CC @gdamore @nawordar

This tree-sitter parser is better maintained (manual parser field names) and is now mentioned officially by the tree-sitter org. 

Perhaps, are you comfortable co-maintaining this with me @nawordar ?

### CODE REVIEW

**Comment by nawordar** [comment]
I'm not using D that much anymore, so I would be fine if I was delisted as a maintainer.

**Comment by theHamsta** [comment]
The old parser took over a minute to compile (probably also had impact on runtime). So a change is welcome. I didn't have time to test the new parser yet.

**Comment by clason** [comment]
As long as the queries look OK, I think that's a strict improvement (even if a breaking change, so please mark it as such) over the old, experimental, parser, so I'd be happy to merge.

File: queries/d/highlights.scm
Comment (theHamsta) [inline queries/d/highlights.scm:23]: `is`/`in` are they literal `in`/`is` in the language? Then perhaps `@keyword.operator`? What about `not_in`

Comment (ljmf00) [reply queries/d/highlights.scm]: Done

File: README.md
Comment (clason) [inline README.md:202]:
```suggestion
- [x] [d](https://github.com/gdamore/tree-sitter-d) (maintained by @ljmf00)
```

File: lua/nvim-treesitter/parsers.lua
Comment (clason) [inline lua/nvim-treesitter/parsers.lua:296]:
```suggestion
  maintainers = { "@ljmf00" },
```

We should take maintainer status seriously; if you are listed, you will get pinged anytime there is an issue with the language -- be it a reported issue, a PR to the queries, or a breaking lockfile update -- and are expected to react in a timely manner.

**Review by clason — CHANGES_REQUESTED** [review]

File: queries/d/highlights.scm
Comment (theHamsta) [inline queries/d/highlights.scm:239]: apparently `(type (identifier) @type)` is missing. I don't see any type highlights and also couldn't find `@type` in this `highlighs.scm`

Comment (theHamsta) [reply queries/d/highlights.scm:239]: strangely a function name in a call is also a `(type (identifier) @type)`

File: lua/nvim-treesitter/parsers.lua
Comment (ljmf00) [reply lua/nvim-treesitter/parsers.lua:296]: Yes, its ok. I contribute a lot to D and nvim is my daily driver. I prefer for someone to directly ping me, because looking at the PRs is not feasible to me, but I have an RSS feed for whenever someone mentions me on github, even without a direct ping, or mentions `tree-sitter-d` or other relevant keywords for the projects I maintain. @gdamore is also my work colleague, so I can possibly cooperate with him if anything is related to the upstream.

Comment (clason) [reply lua/nvim-treesitter/parsers.lua:296]: We need to look into automating that, as it's also not really feasible to manually ping maintainers for the 190+ languages we have...

Comment (gdamore) [reply lua/nvim-treesitter/parsers.lua:296]: I don't use nvim, or I'd probably have done this myself -- thanks @ljmf00 for doing it.

I'm happy to help out.  Especially with the queries - though I'm not aware of the specific nvim selectors that are relevant.

Comment (clason) [reply lua/nvim-treesitter/parsers.lua:296]: They're documented in https://github.com/nvim-treesitter/nvim-treesitter/blob/master/CONTRIBUTING.md (but we'll rework them a bit to be more in line with upstream/Helix in the near future).

Comment (ljmf00) [reply lua/nvim-treesitter/parsers.lua:296]: There's a way to automatically add them via CODEOWNERS and also use github-actions with an heuristic to automatically mention possible maintainers, like based on the tree-sitter link or add a PR/issue template with a dropdown with the language and then map the language to a person.

File: queries/d/highlights.scm
Comment (gdamore) [reply queries/d/highlights.scm:239]: Grammatically in D -- the distinction between a function call and a type is not something you can easily distinguish using syntax alone in many circumstances.   (I don't love D's grammar, but I did the tree-sitter plugin because I'm forced to work in it at work.)

File: queries/d/highlights.scm
Comment (gdamore) [reply queries/d/highlights.scm:212]: The @exception selector is missing in the other editor plugins I've done -- but it would be easy to fix that here by moving those words into a block from where they are located.

Comment (gdamore) [reply queries/d/highlights.scm:212]: Oooh, this looks like an error of omission.

File: lua/nvim-treesitter/parsers.lua
Comment (clason) [reply lua/nvim-treesitter/parsers.lua:296]: CODEOWNERS only works for actual org members, so we'd have to add every (community) maintainer to the team. There are options with github actions, as you're mentioning; we are looking into the different options here (as the list of maintainers itself has to be maintained actively).

Comment (gdamore) [reply lua/nvim-treesitter/parsers.lua:296]: > They're documented in https://github.com/nvim-treesitter/nvim-treesitter/blob/master/CONTRIBUTING.md (but we'll rework them a bit to be more in line with upstream/Helix in the near future).

I wrote the helix plugin for D. So when that happens it will be easier here.  Having said that it's possible that I am missing some there too - I will check on it.

**Comment by clason** [comment]
@ljmf00 ping on the requested changes

File: README.md
Comment (ljmf00) [reply README.md]: Done

File: lua/nvim-treesitter/parsers.lua
Comment (ljmf00) [reply lua/nvim-treesitter/parsers.lua]: Done

File: queries/d/highlights.scm
Comment (ljmf00) [reply queries/d/highlights.scm:212]: Added. Thanks for noticing!

File: lua/nvim-treesitter/parsers.lua
Comment (clason) [inline lua/nvim-treesitter/parsers.lua:294]: If `main` is the default branch, this can be omitted.

File: queries/d/highlights.scm
Comment (ljmf00) [reply queries/d/highlights.scm:245]: Done

File: lua/nvim-treesitter/parsers.lua
Comment (ljmf00) [reply lua/nvim-treesitter/parsers.lua:294]: I had some issues loading it on my nvim without it

Comment (clason) [reply lua/nvim-treesitter/parsers.lua:294]: That doesn't sound like the actual issue; the problem was more likely (since it usually is) the stale lockfile commit hash.

Try dropping it.

File: queries/d/highlights.scm
Comment (ljmf00) [reply queries/d/highlights.scm:245]: I'm still not able to get types highlighted :/

**Comment by ljmf00** [comment]
Don't merge, see https://github.com/nvim-treesitter/nvim-treesitter/pull/4898#discussion_r1230149947 .

**Comment by ljmf00** [comment]
@gdamore I think your some of your highlights queries are wrong, or perhaps `nvim-treesitter` treats queries differently. This seem to override every identifier https://github.com/gdamore/tree-sitter-d/blob/main/queries/helix-highlights.scm#L230 .

@clason Look's like it's working correctly now:
image
Source code from DMD (`expression.d` file)

**Comment by clason** [comment]
Yes, nvim-treesitter handles query order differently than upstream (last query wins, since all are applied in order, rather than first matching query short-circuits).

**Comment by ljmf00** [comment]
> Yes, nvim-treesitter handles query order differently than upstream (last query wins, since all are applied in order, rather than first matching query short-circuits).

Is there any reason to do so? It seems to me that short circuiting is more efficient to me.

**Comment by clason** [comment]
The reason is that this is easier to implement (and hence more efficient) and easier to reason about. Even upstream agrees now that this is better.

**Comment by ljmf00** [comment]
Just squash to `feat!()` to mark as breaking change.

**Review by clason — APPROVED** [review]

**Comment by amaanq** [comment]
@ljmf00 I made a lot of changes locally to the highlights to make it better (a bit too many to suggest changes here), would you mind if I force push those changes here for you to try out and review?

**Comment by clason** [comment]
ping @ljmf00 are you willing to make the changes needed to avoid the regressions?

**Comment by clason** [comment]
That's fine, there's no hurry, I just wanted to check whether you were still interested in this. (But if you are so busy, you probably want to look for a second maintainer as we do expect a reasonable reaction time in case of breakages.)

**Comment by ljmf00** [comment]
> That's fine, there's no hurry, I just wanted to check whether you were still interested in this. (But if you are so busy, you probably want to look for a second maintainer as we do expect a reasonable reaction time in case of breakages.)

I am. I use this everyday on my nvim, as I develop in D including contributions to D language itself, but my time has been very limited, and I suppose my colleague @gdamore , who wrote this grammar implementation, too. If you want to take it @gdamore , just suggest some superseded PR and tell to drop this one. I added this to my personal task list and prioritize it, but it's been difficult to have some time for it.

**Comment by clason** [comment]
Superseded by #5988 

Thanks for getting this started!

**Pull request closed without merging**
